asgiref/typing.py

```diff
@@ -78,12 +94,14 @@ class WebSocketScope(TypedDict):
     client: Optional[Tuple[str, int]]
     server: Optional[Tuple[str, Optional[int]]]
     subprotocols: Iterable[str]
+    state: NotRequired[Dict[str, Any]]
     extensions: Optional[Dict[str, Dict[object, object]]]
 
 
 class LifespanScope(TypedDict):
     type: Literal["lifespan"]
     asgi: ASGIVersions
+    state: NotRequired[Dict[str, Any]]
```

Comment:
@andrewgodwin gave this a last pass and noticed that this was `Optional[Dict[str, Any]]` but I think it should be `NotRequired[Dict[str, Any]]` for the same reasons as https://github.com/django/asgiref/pull/354/files#r1118125962 so I changed it. Just an FYI since you already reviewed.

Reply:
I agree - I also missed that!

Reply:
@adriangb why can't this be `Dict[Any, Any]` rather than `Dict[str, Any]`?

Below, the docs state: 
> Applications can store arbitrary data in this namespace.

Why wouldn't that include arbitrary keys?

> applications should be cooperative by properly naming their keys such that they
will not collide with other frameworks or middleware.

The most foul-proof way to avoid collisions would be to not use string keys at all, but objects (similar to `Symbol` keys in javascript).

Also, that would help make dependency injection simpler in FastAPI/Starlette.
Example use-case:

```python
# (roughly)
HTTP_CLIENT: ClientSession = fastapi.Depends(lambda request: request.scope['state'][HTTP_CLIENT])
DB_POOL: Pool = fastapi.Depends(lambda request: request.scope['state'][DB_POOL])

@asynccontextmanager
async def app_lifespan(app: FastAPI):
    async with make_http_client() as client:
        async with make_database_connection_pool() as db_pool:
            yield {HTTP_CLIENT: client, DB_POOL: db_pool}
```

I am already doing this and it works just fine! However, it is annoying that the typing doesn't match up quite right... and I don't see a reason keys should be limited to strings.

Reply:
Starlette already assumes string keys for its state object, so I kind of copied that here.
It's always possible to do `scope["state"]["mykey"] = {object(): 123}`.
Finally, I think if FastAPI ends up wanting to use this feature and shows a compelling use case for making the keys `Any` we can re-evaluate and change the typing.

Reply:
@adriangb thanks for the response. This isn't something I'd think would be part of the fastapi codebase (although, maybe, who knows). This is more a user-land example of a use-case for non-string keys. Would it be more appropriate for me to raise a similar issue in Starlette or Uvicorn then? As I mentioned, I am already doing the above approach (nothing prevents me). It's just that the typing is a lie, since I'm storing non-string keys. Since Starlette don't have any control over what kinds of keys are returned from the app lifespan function (example above), and similarly, Uvicorn doesn't have any control over what keys its lifespan state dict is updated with, it doesn't make sense to me to pretend they are definitely strings.

Reply:
One thing in particular that will break is Starlette's `State` class: https://github.com/encode/starlette/blob/15d9350eed9437bcb0683e2c24cd4cfbf985ed7e/starlette/datastructures.py#L683-L708. Now, I'm not a fan of that class/functionality, but it exists. Of course, if you don't care you can continue doing what you are doing, but then solving the typing issue is also as simple as a `type: ignore`.
I'm also curious why the suggestion above of something like `yield {"dependencies": {...}}` or something like that wouldn't work.

All of that said, if you feel strongly please feel free to open an issue / PR so it can be discussed properly.